tests/components/mqtt/test_init.py

```diff
@@ -1053,14 +1112,134 @@ async def test_restore_all_active_subscriptions_on_reconnect(
+async def test_initial_setup_logs_error(hass, caplog, mqtt_client_mock):
+    """Test for setup failure if initial client connection fails."""
+    entry = MockConfigEntry(domain=mqtt.DOMAIN, data={mqtt.CONF_BROKER: "test-broker"})
+
+    mqtt_client_mock.connect.return_value = 1
+    assert await mqtt.async_setup_entry(hass, entry)
+    await hass.async_block_till_done()
+    assert "Failed to connect to MQTT server:" in caplog.text
+
+
+async def test_logs_error_if_no_connect_broker(
+    hass, caplog, mqtt_mock, mqtt_client_mock
+):
     """Test for setup failure if connection to broker is missing."""
+    # test with rc = 3 -> broker unavailable
+    mqtt_client_mock.on_connect(mqtt_client_mock, None, None, 3)
+    await hass.async_block_till_done()
+    assert (
+        "Unable to connect to the MQTT broker: Connection Refused: broker unavailable."
+        in caplog.text
+    )
+
+
+@patch("homeassistant.components.mqtt.TIMEOUT_ACK", 0.3)
+async def test_handle_mqtt_on_callback(hass, caplog, mqtt_mock, mqtt_client_mock):
+    """Test receiving an ACK callback before waiting for it."""
+    # Simulate an ACK for mid == 1, this will call mqtt_mock._mqtt_handle_mid(mid)
+    mqtt_client_mock.on_publish(mqtt_client_mock, None, 1)
+    await hass.async_block_till_done()
+    # Make sure the ACK has been received
+    await hass.async_block_till_done()
+    # Now call publish without call back, this will call _wait_for_mid(msg_info.mid)
+    await mqtt.async_publish(hass, "no_callback/test-topic", "test-payload")
+    # Since the mid event was already set, we should not see any timeout
+    await hass.async_block_till_done()
+    assert (
+        "Transmitting message on no_callback/test-topic: 'test-payload', mid: 1"
+        in caplog.text
+    )
+    assert "No ACK from MQTT server" not in caplog.text
+
+
+async def test_publish_error(hass, caplog):
+    """Test publish error."""
     entry = MockConfigEntry(domain=mqtt.DOMAIN, data={mqtt.CONF_BROKER: "test-broker"})
 
+    # simulate an Out of memory error
     with patch("paho.mqtt.client.Client") as mock_client:
         mock_client().connect = lambda *args: 1
+        mock_client().publish().rc = 1
+        assert await mqtt.async_setup_entry(hass, entry)
+        await hass.async_block_till_done()
+        with pytest.raises(HomeAssistantError):
+            await mqtt.async_publish(
+                hass, "some-topic", b"test-payload", qos=0, retain=False, encoding=None
+            )
+        assert "Failed to connect to MQTT server: Out of memory." in caplog.text
+
+
+async def test_handle_message_callback(hass, caplog, mqtt_mock, mqtt_client_mock):
+    """Test for handling an incoming message callback."""
+    msg = ReceiveMessage("some-topic", b"test-payload", 0, False)
+    mqtt_mock().connected = True
+    await mqtt.async_subscribe(hass, "some-topic", lambda *args: 0)
+    mqtt_mock._mqtt_on_message(mock_mqtt, None, msg)
```

Comment:
Preferably we should call the client callback here and not the integration callback.

Reply:
done, added comment to indicate we want to test calls through _mqtt_on_message

Reply:
This line wasn't changed.

Reply:
I'll do that tomorrow